Simplify managed MetadataImport creation

AaronRobinsonMSFT opened the pull request:
Consolidate the create mechanism for the managed
`MetadataImport` type. This also removes some
Helper Method Frame usage.

This is follow-up to work done in https://github.com/dotnet/runtime/pull/100939.

AaronRobinsonMSFT committed:
Simplify managed MetadataImport creation

Consolidate the create mechanism for the managed
MetadataImport type. This also removes some
Helper Method Frame usage.

AaronRobinsonMSFT commented:
/cc @jkotas @jkoritzinsky

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs:
```diff
@@ -650,12 +650,9 @@ internal static bool SatisfiesConstraints(RuntimeType paramType, RuntimeType[]?
         internal static MetadataImport GetMetadataImport(RuntimeType type)
         {
-            return new MetadataImport(_GetMetadataImport(type), type);
+            return new MetadataImport(type.GetRuntimeModule(), type);
```
This is slower than the existing code and it can trigger allocation of the RuntimeModule instance when there was no allocation before. Unclear whether it is problem in practice.

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/MdImport.cs:
```diff
@@ -304,9 +304,19 @@ internal static unsafe MarshalAsAttribute GetMarshalAs(ConstArray nativeType, Ru
+        {
+            ArgumentNullException.ThrowIfNull(module);
+            m_metadataImport2 = GetMetadataImport(module);
             m_keepalive = keepalive;
```
The handling of keepalive is questionable. The JIT can inline this constructor, observe that `m_keepalive` is unused, and optimize it out so it won't serve the intended purpose.

We may be lucky today that it is not causing any observable crash. Harmless refactorings like this one may turn this into observable crash.

We may want to refactor this into something more robust. For example:
- Delete keepalive here
- Add IntPtr m_metadataImport2 as a field to `RuntimeModule`. It would address my other comment about potential perf impact.
- Make sure that all places that use m_metadataImport2 keep the RuntimeModule alive for as long as it is used (e.g. add GC.KeepAlive at the end of the method, we have many places like that)

AaronRobinsonMSFT committed:
Remove MetadataImport.EmptyImport

AaronRobinsonMSFT committed:
Cast default appropriately.

AaronRobinsonMSFT committed:
Remove keepalive member.

jkotas reviewed (approved):
Thanks!

AaronRobinsonMSFT committed:
Update src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

AaronRobinsonMSFT committed:
Always get the latest metadataimport instance.
The field on the managed object can become
stale during HotReload or EnC scenarios.
